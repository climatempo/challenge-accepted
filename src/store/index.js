import { createStore, applyMiddleware } from 'redux';
import ReduxThunk from 'redux-thunk';
import reducers from './reducers';

const devTools = window.__REDUX_DEVTOOLS_EXTENSION__ && window.__REDUX_DEVTOOLS_EXTENSION__();

const store = applyMiddleware(ReduxThunk)(createStore)(reducers, devTools);

export default store;
