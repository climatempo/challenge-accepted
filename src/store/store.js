import { createStore, applyMiddleware, compose } from "redux";
import thunk from 'redux-thunk';

import reducer from "../reducer/reducer";

export default createStore(reducer, compose(applyMiddleware(thunk)));