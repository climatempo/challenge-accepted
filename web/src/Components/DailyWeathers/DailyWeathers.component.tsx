import { useAppSelector, useServerAddress } from '../../hooks';
import { useEffect, useState } from 'react';
import { CircularProgress, Divider, Typography } from '@mui/material';
import SubInfo from '../SubInfo';
import { useSnackbar } from 'notistack';
// @ts-ignore
import { WiThermometer, WiSunset, WiStrongWind, WiSunrise, WiWindDeg, WiRaindrop, WiRaindrops, WiUmbrella, WiThermometerExterior } from 'weather-icons-react';

type DailyWeather = {
    idcity: number,
    moon: string[],
    rainbow: string,
    description: string,
    date: string,
    temperature: {
        min: number,
        max: number
    },
    rain: {
        precipitation: number,
        probability: number
    },
    wind: {
        compass: string,
        velocity: number
    },
    humidity: {
        min: number,
        max: number
    },
    sun: {
        morning: string,
        afternoon: string
    }
}

function DailyWeathers() {
    const locale = useAppSelector(state => state.locale.value);
    const tempUnity = useAppSelector(state => state.temperature.value);
    const rainUnity = useAppSelector(state => state.precipitation.value);
    const [weathers, setWeathers] = useState<DailyWeather[]>([]);
    const [loading, setLoading] = useState<boolean>(true);
    const { enqueueSnackbar: notify } = useSnackbar();
    const server = useServerAddress();

    useEffect(() => {
        if(!locale)
            return;
        setLoading(true);
        fetch(`${ server }/city/${ locale.idcity }/weathers`)
            .then(response => {
                setLoading(false);
                if(response.status === 404) {
                    notify('Não temos o clima dessa cidade.', { variant: 'info'});
                    return [];
                }else return response.json();
            })
            .then(weathers => setWeathers(weathers))
            .catch(err => {
                console.error(err);
                setLoading(false);
                notify('Alguma coisa deu errado.', { variant: 'error'});
            });
    }, [locale, notify, server]);

    function renderDate(weather: DailyWeather) {
        const date = weather.date.split('-');

        return <div className="flex gap-1 items-baseline mt-[-1rem] flex-shrink-0">
            <Typography variant="h1">
                { date[2] }
            </Typography>
            <Typography variant="h5" className="flex-shrink-0">
                { '/ '+date[1] }
            </Typography>
        </div>;
    }

    function renderDailyWeatherCard(weather: DailyWeather, key: number) {
        const {
            description,
            temperature: { min: tmin, max: tmax },
            rain: { precipitation, probability },
            humidity: { min: hmin, max: hmax },
            wind: { compass, velocity },
            rainbow,
            moon,
            sun: { morning, afternoon }
        } = weather;
        const moons = moon.map(moon => moon.split(' - '));

        return <div  key={key}>
            <Divider variant="fullWidth" />
            <div className="flex gap-8 my-20">
            <div className="hidden lg:block">
                { renderDate(weather) }
            </div>
            <div className="flex flex-col items-baseline w-full">
                <div className="flex flex-col sm:flex-row gap-0 sm:gap-8">
                    <div className="block lg:hidden">
                        { renderDate(weather) }
                    </div>
                    <div>
                        <Typography variant="h6">
                            { description }
                        </Typography>
                        <div className="flex flex-col lg:flex-row lg:gap-8 mb-4">
                            <Typography variant="overline">{ rainbow } de arco-íris</Typography>
                            <Typography variant="overline">
                                { moon.length === 1 ? 'Lua '+moon[0] : `Lua ${moons[0][0]} até ${moons[0][1]} e ${moons[1][0]} após ${moons[1][1]}`}
                            </Typography>

                        </div>
                    </div>
                </div>
                <div className="flex flex-wrap justify-start gap-6 sm:gap-0 sm:justify-between w-full">
                    <SubInfo bottomIcon={<WiThermometerExterior size={28} className="mr-2"/>}
                             bottomValue={tempUnity === 'c' ? tmin : ((tmin * 1.8) + 32).toFixed(0)}
                             bottomCaption={<div className="mt-[-2.5rem]">{tempUnity === 'c' ? '°C' : '°F'}</div>}
                             topIcon={<WiThermometer size={28} className="mr-2"/>}
                             topValue={tempUnity === 'c' ? tmax : ((tmax * 1.8) + 32).toFixed(0)}
                             topCaption={<div className="mt-[-2.5rem]">{tempUnity === 'c' ? '°C' : '°F'}</div>}
                             caption="Temperatura"/>
                    <SubInfo bottomIcon={<WiRaindrops size={28} className="mr-2"/>}
                             bottomValue={rainUnity === 'mm' ? precipitation : (precipitation / 25.4).toFixed(1)}
                             bottomCaption={rainUnity === 'mm' ? 'mm' : 'inch'}
                             topIcon={<WiUmbrella size={28} className="mr-2"/>}
                             topValue={probability}
                             topCaption={<div className="mt-[-2.5rem]">%</div>}
                             caption="Chuva"/>
                    <SubInfo bottomIcon={<WiRaindrops size={28} className="mr-2"/>}
                             bottomValue={hmin}
                             bottomCaption={<div className="mt-[-2.5rem]">%</div>}
                             topIcon={<WiRaindrop size={28} className="mr-2"/>}
                             topValue={hmax}
                             topCaption={<div className="mt-[-2.5rem]">%</div>}
                             caption="Humidade"/>
                    <SubInfo bottomIcon={<WiWindDeg size={28} className="mr-2"/>}
                             bottomValue={compass}
                             topIcon={<WiStrongWind size={28} className="mr-2"/>}
                             topValue={tempUnity === 'c' ? velocity : (velocity/1.609).toFixed(0)}
                             topCaption={tempUnity === 'c' ? 'km/h' : 'mph'}
                             caption="Vento"/>
                    <SubInfo bottomIcon={<WiSunrise size={28} className="mr-2"/>}
                             bottomValue={morning}
                             topIcon={<WiSunset size={28} className="mr-2"/>}
                             topValue={afternoon}
                             caption="Sol"/>
                </div>
            </div>
        </div></div>;
    }

    function renderSkeleton() {
        return loading ? <CircularProgress className="mx-auto" /> : <></>
    }
    
    return <div className="mt-24 flex flex-col">
        { weathers.length && !loading ? weathers.map(renderDailyWeatherCard) : renderSkeleton() }
    </div>;
}

export default DailyWeathers;