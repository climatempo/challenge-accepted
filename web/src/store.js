import Vue from "vue";
import Vuex from "vuex";
import localeStore from "./store/locale";
import weatherStore from "./store/weather";

Vue.use(Vuex);

export default new Vuex.Store({
  state: {},
  mutations: {},
  actions: {},
  modules: [
      localeStore,
      weatherStore
  ]
});
