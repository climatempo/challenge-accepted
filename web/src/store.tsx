import { configureStore } from '@reduxjs/toolkit';
import themeReducer from './Reducers/Theme.reducer';
import temperatureReducer from './Reducers/Temperature.reducer';
import precipitationReducer from './Reducers/Precipitation.reducer';

const store = configureStore({
    reducer: {
        theme: themeReducer,
        temperature: temperatureReducer,
        precipitation: precipitationReducer,
    }
});

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;
export default store;