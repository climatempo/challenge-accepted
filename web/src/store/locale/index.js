import mutations from './mutations'
import actions from './actions'
import getters from './getters'

/**
 * Store de localizaçeõs
 *
 * @type mixins
 */
const localeStore = {
    state: {
        locales: []
    },
    mutations: mutations,
    actions: actions,
    getters: getters
};

export default localeStore
