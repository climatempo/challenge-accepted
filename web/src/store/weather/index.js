import mutations from './mutations'
import actions from './actions'
import getters from './getters'

/**
 * Store de previs√µes
 *
 * @type mixins
 */
const weatherStore = {
    state: {
        weather: []
    },
    mutations: mutations,
    actions: actions,
    getters: getters
};

export default weatherStore
