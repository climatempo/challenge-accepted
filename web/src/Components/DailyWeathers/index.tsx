import DailyWeathers from './DailyWeathers.component';

export default DailyWeathers;