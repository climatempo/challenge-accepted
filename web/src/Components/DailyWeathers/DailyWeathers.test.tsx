import renderer from 'react-test-renderer';
import './../../Tests/watchMedia.mock';
import DailyWeathers from './index';
import { Provider } from 'react-redux';
import store from '../../store';
import { SnackbarProvider } from 'notistack';

test("Daily Weather component", () => {
  const reduxStore = store;

  const weathers = renderer.create(
      <Provider store={reduxStore}>
        <SnackbarProvider maxSnack={4}>
          <DailyWeathers/>
        </SnackbarProvider>
      </Provider>
  );

  expect(weathers.toJSON()).toMatchSnapshot();
});