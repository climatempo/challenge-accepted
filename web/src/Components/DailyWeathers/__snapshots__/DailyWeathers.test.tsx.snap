// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Daily WeatherController component 1`] = `
<div
  className="mt-24 flex flex-col"
>
  <span
    className="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorPrimary mx-auto css-18lrjg1-MuiCircularProgress-root"
    role="progressbar"
    style={
      Object {
        "height": 40,
        "width": 40,
      }
    }
  >
    <svg
      className="MuiCircularProgress-svg css-1idz92c-MuiCircularProgress-svg"
      viewBox="22 22 44 44"
    >
      <circle
        className="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-176wh8e-MuiCircularProgress-circle"
        cx={44}
        cy={44}
        fill="none"
        r={20.2}
        strokeWidth={3.6}
        style={Object {}}
      />
    </svg>
  </span>
</div>
`;
